ziti/cmd/ziti/cmd/edge_controller/login.go

```diff
@@ -80,70 +81,199 @@ func newLoginCmd(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Comm
-		return errors.Errorf("could not parse host, invalid: %v", err)
+	if err = o.ConfigureCerts(host, ctrlUrl); err != nil {
+		return err
 	}
 
-	if hostUrl.Path == "" {
+	if certAbs, err := filepath.Abs(o.Cert); err == nil {
+		o.Cert = certAbs
+	}
+
+	if ctrlUrl.Path == "" {
 		host = util.EdgeControllerGetManagementApiBasePath(host, o.Cert)
 	}
 
-	if o.Username != "" && o.Password == "" {
-		var err error
-		if o.Password, err = term.PromptPassword("Enter password: ", false); err != nil {
+	if o.Token == "" {
+		for o.Username == "" {
+			if o.Username, err = term.Prompt("Enter username: "); err != nil {
+				return err
+			}
+		}
+
+		if o.Username != "" && o.Password == "" {
+			if o.Password, err = term.PromptPassword("Enter password: ", false); err != nil {
+				return err
+			}
+		}
+
+		container := gabs.New()
+		_, _ = container.SetP(o.Username, "username")
+		_, _ = container.SetP(o.Password, "password")
+
+		body := container.String()
+
+		jsonParsed, err := util.EdgeControllerLogin(host, o.Cert, body, o.Out, o.OutputJSONResponse, o.edgeOptions.Timeout, o.edgeOptions.Verbose)
+
+		if err != nil {
 			return err
 		}
+
+		if !jsonParsed.ExistsP("data.token") {
+			return fmt.Errorf("no session token returned from login request to %v. Received: %v", host, jsonParsed.String())
+		}
+
+		var ok bool
+		o.Token, ok = jsonParsed.Path("data.token").Data().(string)
+
+		if !ok {
+			return fmt.Errorf("session token returned from login request to %v is not in the expected format. Received: %v", host, jsonParsed.String())
+		}
+
+		if !o.OutputJSONResponse {
+			o.Printf("Token: %v\n", o.Token)
+		}
 	}
-	container := gabs.New()
-	_, _ = container.SetP(o.Username, "username")
-	_, _ = container.SetP(o.Password, "password")
 
-	body := container.String()
+	absCertPath, err := filepath.Abs(o.Cert)
+	if err == nil {
+		o.Cert = absCertPath
+	}
 
-	jsonParsed, err := util.EdgeControllerLogin(host, o.Cert, body, o.Out, o.OutputJSONResponse, o.edgeOptions.Timeout, o.edgeOptions.Verbose)
+	loginIdentity := &util.RestClientIdentity{
+		Url:       host,
+		Token:     o.Token,
+		LoginTime: time.Now().Format(time.RFC3339),
+		Cert:      o.Cert,
+		ReadOnly:  o.ReadOnly,
+	}
 
+	config, configFile, err := util.LoadRestClientConfig()
 	if err != nil {
 		return err
 	}
 
-	if !jsonParsed.ExistsP("data.token") {
-		return fmt.Errorf("no session token returned from login request to %v. Received: %v", host, jsonParsed.String())
+	id := config.GetIdentity()
+	o.Printf("Saving identity '%v' to %v\n", id, configFile)
+	config.Identities[id] = loginIdentity
+
+	err = util.PersistRestClientConfig(config)
+
+	return err
+}
+
+func (o *loginOptions) ConfigureCerts(host string, ctrlUrl *url.URL) error {
+	isServerTrusted, err := util.IsServerTrusted(host)
+	if err != nil {
+		return err
 	}
 
-	token, ok := jsonParsed.Path("data.token").Data().(string)
+	if isServerTrusted && o.Cert == "" {
+		wellKnownCerts, certs, err := util.GetWellKnownCerts(host)
```

Comment:
After we retrieve them, we should make sure the bundle actually validates the server's certificate.

Reply:
Think that should be straight forward to add

Reply:
Fixed: Added verify after retrieval